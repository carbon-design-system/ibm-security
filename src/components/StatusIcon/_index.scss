////
/// Status icon component.
/// @group status-icon
/// @copyright IBM Security 2019
////

@import '../Loading/index';

@import '../Component/mixins';
@import 'mixins';

@include security--component($name: status-icon) {
  @include status-icon;
}
