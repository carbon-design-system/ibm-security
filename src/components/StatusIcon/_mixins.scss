////
/// Status icon mixins.
/// @group status-icon
/// @copyright IBM Security 2019
////

@import '../../globals/index';

@mixin status-icon {
  /// Success color.
  /// @type Color
  $status-icon--success__color: $support-02;

  display: flex;
  align-items: center;

  &__icon {
    &--success {
      /// Success transform.
      /// @type Length
      $status-icon--success__transform: 0.55;

      padding: $carbon--spacing-03;
      fill: $icon-01;
      border: carbon--mini-units($count: 0.25) solid
        $status-icon--success__color;
      border-radius: 100%;
      box-sizing: border-box;
      transform: scale3d(
        $status-icon--success__transform,
        $status-icon--success__transform,
        $status-icon--success__transform
      );
    }

    &--color {
      /// Color icon dimensions.
      /// @type Color
      $status-icon__icon--color__sizing__dimensions: carbon--mini-units(
        $count: 1,
      );

      /// Types.
      /// @type Map<String, Color>
      $status-icon__icon__types: (
        error: $support-01,
        info: $support-04,
        success: $status-icon--success__color,
        warning: $support-03,
      );

      width: $status-icon__icon--color__sizing__dimensions;
      height: $status-icon__icon--color__sizing__dimensions;
      background-color: $icon-02;
      border-radius: 50%;

      @each $status-icon__icon__type,
        $status-icon__icon__color in $status-icon__icon__types
      {
        &--#{$status-icon__icon__type} {
          background-color: $status-icon__icon__color;
        }
      }
    }
  }

  /// Sizes.
  /// @type Map<String, number>
  $status-icon__sizes: (
    lg: 9,
    md: 5,
    sm: 4,
  );

  @each $status-icon__type, $status-icon__value in $status-icon__sizes {
    &--#{$status-icon__type} > &__icon {
      $status-icon__size: carbon--mini-units(
        $count: $status-icon__value,
      );

      width: $status-icon__size;
      height: $status-icon__size;
    }
  }

  &__message {
    margin-left: $carbon--spacing-03;
  }
}
