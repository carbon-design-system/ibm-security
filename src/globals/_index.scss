////
/// Globals entry point.
/// @group entry
/// @copyright IBM Security 2019 - 2020
////

@import 'theme/index';

@import 'variables/index';

@import 'carbon-components/scss/globals/scss/helper-mixins';

@import 'color/index';
@import 'grid/index';
@import 'layout/index';
@import 'motion/index';
@import 'type/index';

@import 'border/index';
@import 'deprecate/index';
@import 'feature-flags/index';
@import 'namespace/index';
@import 'rtl/index';
