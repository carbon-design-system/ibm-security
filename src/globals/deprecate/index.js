/**
 * @file Deprecation warning.
 * @copyright IBM Security 2019 - 2021
 */

import isDevelopment from '../env';

/**
 * Filters, compares, and logs the appropriate properties with a description.
 * @param {string} description The description to use when logging.
 * @param {string} type The property type to log.
 * @param {Array.<string>} componentKeys The component keys to filter.
 * @param {function} condition The condition to filter the component keys.
 * @returns {string} The description and appropriate properties.
 */
const logDeviations = (description, type, componentKeys, filterCondition) =>
  console.log(`${description} ${type}`, componentKeys.filter(filterCondition));

/**
 * Helper for comparing and logging deviations between multiple components.
 * @param {string} carbonComponent The Carbon component to compare.
 * @param {string} securityComponent The Security-specific component to compare.
 */
const compareDeviations = (carbonComponent, securityComponent) => {
  ['defaultProps', 'propTypes'].forEach(type => {
    const carbonComponentKeys = Object.keys(carbonComponent[type]);
    const securityComponentKeys = Object.keys(securityComponent[type]);

    logDeviations(
      'Common',
      type,
      carbonComponentKeys,
      key => securityComponentKeys.indexOf(key) !== -1
    );

    logDeviations(
      'Carbon',
      type,
      carbonComponentKeys,
      key => securityComponentKeys.indexOf(key) === -1
    );

    logDeviations(
      'Security',
      type,
      securityComponentKeys,
      key => carbonComponentKeys.indexOf(key) === -1
    );
  });
};

/**
 * Returns a standardized deprecation warning for development.
 * @param {string} warning - The warning value supplied.
 * @returns {console.trace} A stack trace containing the created deprecation details.
 */
function deprecationWarning(warning) {
  if (isDevelopment()) {
    try {
      throw new Error(
        `Warning: ${warning} and will be removed in an upcoming major release`
      );
    } catch (error) {
      console.warn(error);
    }
  }
}

/**
 * Returns a standardized deprecation warning.
 * @param {string} actual - The actual value supplied.
 * @param {string} expected - The expected value to be supplied.
 * @returns {console.trace} A stack trace containing the created deprecation details.
 */
const deprecate = (actual, expected) =>
  deprecationWarning(
    `'${actual}' has been deprecated in favor of '${expected}'`
  );

/**
 * Returns a standardized component deprecation warning.
 * @param {Object.<string, string>} deprecationWarning - An object containing properties that construct the deprecation warning.
 * @param {string} deprecationWarning.actual - The actual value supplied.
 * @param {string} deprecationWarning.componentName - The component name it was supplied to.
 * @param {string} deprecationWarning.description - The description of the deprecation warning.
 * @param {string} deprecationWarning.expected - The expected value to be supplied.
 * @returns {console.trace} A stack trace containing the created deprecation details.
 */
export default ({ actual, componentName, description, expected }) =>
  deprecationWarning(
    `Deprecated ${description} '${actual}' supplied to '${componentName}', expected '${expected}'`
  );

export { compareDeviations, deprecate };
