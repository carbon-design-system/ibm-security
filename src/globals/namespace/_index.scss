////
/// Assets related to namespacing selectors.
/// @group namespace
/// @copyright IBM Security 2019
////

@import '@carbon/import-once/scss/import-once';

/// The prefix used to namespace all selectors in the CSS.
/// @type String
/// @access private
$namespace: 'security--';

///
/// Appends the namespace for a component to another namespace.
/// @param {String} $component-namespace The component namespace to append to.
/// @param {String} $component-name The component namespace to append.
/// @return {string} The appended component namespace to return.
///
@function append-component-namespace($component-namespace, $component-name) {
  @return '#{$component-namespace}__#{$component-name}';
}

///
/// Creates the namespace for a component using the global namespace.
/// @param {String} $component-name The component namespace to create.
/// @return {String} The component namespace to return.
///
@function get-component-namespace($component-name) {
  @return '.#{$namespace}#{$component-name}';
}

/// Exports the content once to the unique namespace.
/// @param {String} $name The name to namespace.
/// @content The content to export once.
/// @example scss
///   @include export-namespace($name: platform) {
///     body {
///       background-color: $body__color__background;
///     }
///   }
/// @output css
///   body {
///     background-color: #fff;
///   }
///
@mixin export-namespace($name) {
  @include exports($name: #{$namespace}#{$name}) {
    @content;
  }
}

/// Sets the namespace prefix on to the given selector.
/// @param {String} $selector The selector string to namespace.
/// @content The styles for the namespaced component class name.
/// @example scss
///   @include namespace($selector: header) {
///     background-color: $header__color__background;
///   }
/// @output css
///   .security--header {
///     background-color: #fff;
///   }
///
@mixin namespace($selector) {
  #{get-component-namespace($component-name: $selector)} {
    @content;
  }
}
