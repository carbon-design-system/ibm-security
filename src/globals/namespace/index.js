/**
 * @file Namespace.
 * @copyright IBM Security 2018 - 2021
 */

const {
  settings: { prefix: carbonPrefix },
} = require('carbon-components');

const namespace = 'security--';

/**
 * Appends the namespace for a component to another namespace.
 * @param {string} componentNamespace The component namespace to append to.
 * @param {string} componentName The component namespace to append.
 * @returns {string} The appended component namespace to return.
 */
const appendComponentNamespace = (componentNamespace, componentName) =>
  `${componentNamespace}__${componentName}`;

/**
 * Creates the namespace for a component using the global namespace.
 * @param {string} componentName The component namespace to create.
 * @returns {string} The component namespace to return.
 */
const getComponentNamespace = componentName => `${namespace}${componentName}`;

export { appendComponentNamespace, carbonPrefix, getComponentNamespace };
