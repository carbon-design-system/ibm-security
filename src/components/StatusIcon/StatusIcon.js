/**
 * @file Status icon.
 * @copyright IBM Security 2019
 */

import Checkmark20 from '@carbon/icons-react/lib/checkmark/20';

import classnames from 'classnames';
import { oneOf, string } from 'prop-types';
import React, { Component } from 'react';

import { getComponentNamespace } from '../../globals/namespace';

import Icon from '../Icon';
import Loading from '../Loading';

export const namespace = getComponentNamespace('status-icon');

const defaultSize = 'md';
const SIZE = ['lg', defaultSize, 'sm'];

const STATUS = ['complete', 'error', 'info', 'success', 'unknown', 'warning'];

/**
 * Status icon component.
 */
export default class StatusIcon extends Component {
  static propTypes = {
    /** @type {string} Class name. */
    className: string,

    /** @type {string} Message. */
    message: string,

    /** @type {string} Size. */
    size: oneOf(SIZE),

    /** @type {string} Status. */
    status: oneOf(STATUS),

    /** @type {string} icon aria label. */
    iconDescription: string,
  };

  static defaultProps = {
    className: null,
    message: null,
    size: defaultSize,
    status: undefined,
    iconDescription: null,
  };

  static getDerivedStateFromProps({ status }, state) {
    return status && state.status !== status
      ? {
          status,
        }
      : null;
  }

  state = {
    status: this.props.status,
  };

  render() {
    const { className, message, size, iconDescription } = this.props;
    const { status } = this.state;

    let statusIcon;

    switch (status) {
      case STATUS[0]:
        statusIcon = (
          <Icon
            aria-label={
              iconDescription && iconDescription.length ? iconDescription : null
            }
            className={`${namespace}__icon ${namespace}__icon--success`}
            renderIcon={Checkmark20}
          />
        );
        break;

      case undefined:
        statusIcon = (
          <Loading
            aria-label={
              iconDescription && iconDescription.length ? iconDescription : null
            }
            className={`${namespace}__icon`}
            withOverlay={false}
          />
        );
        break;

      default:
        statusIcon = (
          <span
            aria-label={
              iconDescription && iconDescription.length ? iconDescription : null
            }
            className={`${namespace}__icon--color ${namespace}__icon--color--${status}`}
          />
        );
    }

    return (
      <div
        className={classnames(namespace, className, {
          [`${namespace}--${size}`]: size,
        })}
      >
        {statusIcon}
        {message && <span className={` ${namespace}__message`}>{message}</span>}
      </div>
    );
  }
}

export { SIZE, STATUS };
