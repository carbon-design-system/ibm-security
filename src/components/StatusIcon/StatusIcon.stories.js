/**
 * @file Status icon stories.
 * @copyright IBM Security 2019 - 2021
 */

import { select, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';

import { compose, getDisplayName, lifecycle } from 'recompose';
import React from 'react';

import { security } from '../../../.storybook';

import { StatusIcon } from '../..';
import { SIZE, STATUS } from './StatusIcon';

const storyProps = () => ({
  message: text('Label (message)', 'Label'),
  size: select('Size (size)', SIZE, StatusIcon.defaultProps.size),
  iconDescription: text('Icon ARIA label', 'Icon description.. '),
});

const status = STATUS[0];

storiesOf(security('StatusIcon'), module)
  .add('Default', () => <StatusIcon {...storyProps()} />)
  .add('Status', () => (
    <StatusIcon
      {...storyProps()}
      status={select('Status (status)', STATUS, status)}
    />
  ))
  .add('Dynamic', () => {
    let timeout;

    const StatusIconStory = compose(
      lifecycle({
        componentDidMount() {
          timeout = setTimeout(() => this.setState({ status }), 1000);
        },

        componentWillUnmount: () => clearTimeout(timeout),
      })
    )(StatusIcon);

    StatusIconStory.displayName = getDisplayName(StatusIcon);
    StatusIconStory.__docgenInfo = StatusIcon.__docgenInfo;

    return <StatusIconStory {...storyProps()} />;
  });
