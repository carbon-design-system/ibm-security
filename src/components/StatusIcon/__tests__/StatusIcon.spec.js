/**
 * @file Status icon tests.
 * @copyright IBM Security 2019
 */

import { render } from '@testing-library/react';
import React from 'react';
import renderWithinLandmark from '../../../../config/jest/helpers/renderWithinLandmark';

import StatusIcon, { namespace, STATUS, SIZE } from '../StatusIcon';

import { carbonPrefix } from '../../../globals/namespace';

describe('StatusIcon', () => {
  STATUS.forEach(status =>
    test(`should have no Axe or DAP violations when \`status\` is  '${status}'`, async () => {
      const { container } = render(
        <StatusIcon status={status} message="test message" />
      );
      await expect(container).toHaveNoAxeViolations();
      await expect(container).toHaveNoDAPViolations(
        `StatusIcon with ${status} status`
      );
    })
  );

  test('should have no Axe or DAP violations when `status` is  `undefined`', async () => {
    const { container } = renderWithinLandmark(
      <StatusIcon message="test message" />
    );
    await expect(container).toHaveNoAxeViolations();
    await expect(container).toHaveNoDAPViolations(
      'StatusIcon with `undefined` status'
    );
  });

  SIZE.forEach(size =>
    test(`should apply correct class when \`size\` is  '${size}'`, () => {
      render(<StatusIcon size={size} />);
      expect(document.querySelector(`.${namespace}`)).toHaveClass(
        `${namespace}--${size}`
      );
    })
  );

  STATUS.forEach(status => {
    if (status === 'complete') {
      test('should render with correct class and an icon when `status` is `complete`', () => {
        render(<StatusIcon status={status} />);
        expect(document.querySelector(`.${namespace}__icon`)).toHaveClass(
          `${namespace}__icon--success`
        );
      });
    } else {
      test(`should apply correct class when \`status\` is  '${status}'`, () => {
        render(<StatusIcon status={status} />);
        expect(
          document.querySelector(`.${namespace}__icon--color`)
        ).toHaveClass(`${namespace}__icon--color--${status}`);
      });
    }
  });

  test('should render a loading icon when `status` is `undefined`', () => {
    render(<StatusIcon />);
    expect(document.querySelector(`.${carbonPrefix}loading`)).toBeVisible();
  });

  test('should add a custom class', () => {
    render(<StatusIcon className="custom-class" />);
    expect(document.querySelector(`.${namespace}`)).toHaveClass('custom-class');
  });

  test('should add a message', () => {
    const { queryByText } = render(<StatusIcon message="test message" />);
    expect(queryByText(/test message/i)).toBeVisible();
  });

  test('should add an aria label', () => {
    const { queryByLabelText } = render(
      <StatusIcon iconDescription="has aria" message="test message" />
    );
    expect(queryByLabelText(/has aria/i)).toBeVisible();
  });
});
