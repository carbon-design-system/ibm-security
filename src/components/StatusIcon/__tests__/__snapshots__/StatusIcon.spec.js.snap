// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusIcon renders 1`] = `
<div
  className="security--status-icon security--status-icon--md"
>
  <Loading
    active={true}
    className="security--status-icon__icon"
    description="Active loading indicator"
    small={false}
    withOverlay={false}
  />
</div>
`;

exports[`StatusIcon renders the 'complete' variation 1`] = `
<div
  className="security--status-icon security--status-icon--md"
>
  <Icon
    className="security--status-icon__icon security--status-icon__icon--success"
    path={null}
    renderIcon={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    size={20}
    viewBox="0 0 32 32"
  />
  <span
    className=" security--status-icon__message"
  >
    complete
  </span>
</div>
`;

exports[`StatusIcon renders the 'error' variation 1`] = `
<div
  className="security--status-icon security--status-icon--md"
>
  <span
    className="security--status-icon__icon--color security--status-icon__icon--color--error"
  />
  <span
    className=" security--status-icon__message"
  >
    error
  </span>
</div>
`;

exports[`StatusIcon renders the 'info' variation 1`] = `
<div
  className="security--status-icon security--status-icon--md"
>
  <span
    className="security--status-icon__icon--color security--status-icon__icon--color--info"
  />
  <span
    className=" security--status-icon__message"
  >
    info
  </span>
</div>
`;

exports[`StatusIcon renders the 'lg' variation 1`] = `
<div
  className="security--status-icon security--status-icon--lg"
>
  <Loading
    active={true}
    className="security--status-icon__icon"
    description="Active loading indicator"
    small={false}
    withOverlay={false}
  />
</div>
`;

exports[`StatusIcon renders the 'md' variation 1`] = `
<div
  className="security--status-icon security--status-icon--md"
>
  <Loading
    active={true}
    className="security--status-icon__icon"
    description="Active loading indicator"
    small={false}
    withOverlay={false}
  />
</div>
`;

exports[`StatusIcon renders the 'sm' variation 1`] = `
<div
  className="security--status-icon security--status-icon--sm"
>
  <Loading
    active={true}
    className="security--status-icon__icon"
    description="Active loading indicator"
    small={false}
    withOverlay={false}
  />
</div>
`;

exports[`StatusIcon renders the 'success' variation 1`] = `
<div
  className="security--status-icon security--status-icon--md"
>
  <span
    className="security--status-icon__icon--color security--status-icon__icon--color--success"
  />
  <span
    className=" security--status-icon__message"
  >
    success
  </span>
</div>
`;

exports[`StatusIcon renders the 'unknown' variation 1`] = `
<div
  className="security--status-icon security--status-icon--md"
>
  <span
    className="security--status-icon__icon--color security--status-icon__icon--color--unknown"
  />
  <span
    className=" security--status-icon__message"
  >
    unknown
  </span>
</div>
`;

exports[`StatusIcon renders the 'warning' variation 1`] = `
<div
  className="security--status-icon security--status-icon--md"
>
  <span
    className="security--status-icon__icon--color security--status-icon__icon--color--warning"
  />
  <span
    className=" security--status-icon__message"
  >
    warning
  </span>
</div>
`;
